homeassistant/components/reolink/media_source.py

```diff
@@ -0,0 +1,329 @@
+        if _LOGGER.isEnabledFor(logging.DEBUG):
+            url_log = f"{url.split('&user=')[0]}&user=xxxxx&password=xxxxx"
+            _LOGGER.debug(
+                "Opening VOD stream from %s: %s", host.api.camera_name(channel), url_log
+            )
+
+        stream = create_stream(self.hass, url, {}, DynamicStreamSettings())
+        stream.add_provider("hls", timeout=3600)
+        stream_url: str = stream.endpoint_url("hls")
+        stream_url = stream_url.replace("master_", "")
+        return PlayMedia(stream_url, mime_type)
+
+    async def async_browse_media(
+        self,
+        item: MediaSourceItem,
+    ) -> BrowseMediaSource:
+        """Return media."""
+        if item.identifier is None:
+            return await self._async_generate_root()
+
+        identifier = item.identifier.split("|", 7)
+        item_type = identifier[0]
+
+        if item_type == "CAM":
+            _, config_entry_id, channel_str = identifier
+            return await self._async_generate_resolution_select(
+                config_entry_id, int(channel_str)
+            )
+        if item_type == "RES":
+            _, config_entry_id, channel_str, stream = identifier
+            return await self._async_generate_camera_days(
+                config_entry_id, int(channel_str), stream
+            )
+        if item_type == "DAY":
+            (
+                _,
+                config_entry_id,
+                channel_str,
+                stream,
+                year_str,
+                month_str,
+                day_str,
+            ) = identifier
+            return await self._async_generate_camera_files(
+                config_entry_id,
+                int(channel_str),
+                stream,
+                int(year_str),
+                int(month_str),
+                int(day_str),
+            )
+
+        raise Unresolvable(f"Unknown media item '{item.identifier}' during browsing.")
+
+    async def _async_generate_root(self) -> BrowseMediaSource:
+        """Return all available reolink cameras as root browsing structure."""
+        children: list[BrowseMediaSource] = []
+
+        entity_reg = er.async_get(self.hass)
+        device_reg = dr.async_get(self.hass)
+        for config_entry in self.hass.config_entries.async_entries(DOMAIN):
+            if config_entry.state != ConfigEntryState.LOADED:
+                continue
+            channels: list[str] = []
+            host = self.data[config_entry.entry_id].host
+            entities = er.async_entries_for_config_entry(
+                entity_reg, config_entry.entry_id
+            )
+            for entity in entities:
+                if (
+                    entity.disabled
+                    or entity.device_id is None
+                    or entity.domain != CAM_DOMAIN
+                ):
+                    continue
+
+                device = device_reg.async_get(entity.device_id)
+                ch = entity.unique_id.split("_")[1]
```

Comment:
`partition` is faster than split if you can use it here

Reply:
thanks, done

Reply:
Actually this is not possible since the unique_id = `mac_channel_key` , so for instance `abcdefghij_1_sub` and `abcdefghij_1_main`. I need only the channel.